pkg/agent/nodeportlocal/portcache/port_table_others.go

```diff
@@ -128,11 +93,38 @@ func (pt *PortTable) AddRule(podIP string, podPort int, protocol string) (int, e
 		pt.addPortTableCache(npData)
 	} else {
 		// Only add rules if the entry does not exist.
-		return 0, fmt.Errorf("existed Linux Nodeport entry for %s:%d:%s", podIP, podPort, protocol)
+		return 0, fmt.Errorf("existing Linux Nodeport entry for %s:%d:%s", podIP, podPort, protocol)
 	}
 	return npData.NodePort, nil
 }
 
+func (pt *PortTable) deleteRule(data *NodePortData) error {
+	protocolSocketData := &data.Protocol
+	protocol := protocolSocketData.Protocol
+
+	// In theory, we should not be modifying a cache item in-place. However, the field we are
+	// modifying (defunct) does NOT participate in indexing and the modification is thread-safe
+	// because of pt.tableLock.
+	// TODO: stop modifying cache items in-place.
+	// We could set defunct after the call to DeleteRule, because a failed call to DeleteRule
+	// should mean that the rule is still present and valid, but there is no harm in being more
+	// conservative.
+	data.defunct = true
```

Comment:
Should we update `defunct` after PodPortRules.DeleteRule executes successfully?

Reply:
We should definitely retry if iptables deletion (call to `pt.PodPortRules.DeleteRule`) fails.
All of these versions are acceptable, but there is no real difference between them IMO:

```go
data.defunct = true
if err := pt.PodPortRules.DeleteRule(data.NodePort, data.PodIP, data.PodPort, protocol); err != nil {
		return err
}
```

```go
if err := pt.PodPortRules.DeleteRule(data.NodePort, data.PodIP, data.PodPort, protocol); err != nil {
		return err
}
data.defunct = true
```

```go
if !data.defunct {
		if err := pt.PodPortRules.DeleteRule(data.NodePort, data.PodIP, data.PodPort, protocol); err != nil {
				return err
		}
		data.defunct = true
}
```

Note that if `pt.PodPortRules.DeleteRule` fails, we will always retry no matter what. We return an error, and so the Pod will be re-queued.
The key change is really to ensure that a partially deleted rule cannot be re-used just because it has been kept in the port table cache. So the one thing we need to guarantee is that `partial deletion => defunct == true`.

if `pt.PodPortRules.DeleteRule` really never succeeds for a rule, then we will always get into an endless error loop. Your second option doesn't change that, unless I am missing something. But we do assume (in this controller and in other agent controllers) that iptables rule deletion will eventually succeed.

Reply:
I see, this error is returned by syncPod in both cases. Thanks for the explanation！